File: app/src/main/java/me/tatarka/betterfragment/sample/MainActivity.java
package me.tatarka.betterfragment.sample;

import android.content.Intent;
import android.content.res.Configuration;
import android.os.Bundle;

import com.google.android.material.bottomnavigation.BottomNavigationView;

import javax.inject.Inject;

import androidx.annotation.Nullable;
import me.tatarka.betterfragment.app.Fragment;
import me.tatarka.betterfragment.appcompat.app.AppCompatActivity;
import me.tatarka.betterfragment.host.FragmentPageHostUI;
import me.tatarka.betterfragment.sample.dagger.DaggerAppComponent;
import me.tatarka.betterfragment.sample.dagger.DaggerFragmentFactory;
import me.tatarka.betterfragment.widget.FragmentPageHost;

public class MainActivity extends AppCompatActivity {

    @Inject
    DaggerFragmentFactory fragmentFactory;

    private FragmentPageHost pageHost;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        DaggerAppComponent.create().inject(this);

        setContentView(R.layout.activity_main);
        pageHost = findViewById(R.id.page_host);
        BottomNavigationView bottomNav = findViewById(R.id.bottom_nav);
        pageHost.setAdapter(new FragmentPageHost.Adapter() {
            @Nullable
            @Override
            public Fragment newInstance(int id) {
                switch (id) {
                    case R.id.simple_host:
                        return fragmentFactory.newInstance(SimpleHostFragment.class);
                    case R.id.view_pager:
                        return fragmentFactory.newInstance(ViewPagerFragment.class);
                    case R.id.navigation:
                        return fragmentFactory.newInstance(NavigationFragment.class);
                }
                return null;
            }
        });
        FragmentPageHostUI.setupWithBottomNavigationView(pageHost, bottomNav);
    }

    @Override
    public void onBackPressed() {
        Fragment fragment = pageHost.getFragment();
        if (!(fragment instanceof NavInterface && ((NavInterface) fragment).onBackPressed())) {
            super.onBackPressed();
        }
    }

    @Override
    public boolean onNavigateUp() {
        Fragment fragment = pageHost.getFragment();
        if (fragment instanceof NavInterface && ((NavInterface) fragment).onNavigateUp()) {
            return true;
        } else {
            return super.onNavigateUp();
        }
    }
}

File: app/src/main/java/me/tatarka/shard/sample/MainActivity.kt
package me.tatarka.shard.sample

import android.os.Bundle
import androidx.lifecycle.get
import com.google.android.material.bottomnavigation.BottomNavigationView
import me.tatarka.shard.appcompat.app.AppCompatActivity
import me.tatarka.shard.lifecycle.ViewModelProviders
import me.tatarka.shard.sample.dagger.injector
import me.tatarka.shard.wiget.ShardPageHost

class MainActivity : AppCompatActivity() {

    lateinit var pageHost: ShardPageHost

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val shardFactory = injector.shardFactory

        ViewModelProviders.of(this).get<MyViewModel>()

        setContentView(R.layout.activity_main)
        pageHost = findViewById(R.id.page_host)
        val bottomNav: BottomNavigationView = findViewById(R.id.bottom_nav)!!
        pageHost.adapter = ShardPageHost.Adapter { id ->
            when (id) {
                R.id.simple_host -> shardFactory.newInstance<SimpleHostShard>()
                R.id.view_pager -> shardFactory.newInstance<ViewPagerShard>()
                R.id.navigation -> shardFactory.newInstance<NavigationShard>()
                R.id.dialogs -> shardFactory.newInstance<DialogHostShard>()
                else -> null
            }
        }
        bottomNav.setupWithPageHost(pageHost)
    }

    override fun onBackPressed() {
        val shard = pageHost.shard
        if (!(shard is NavInterface && shard.onBackPressed())) {
            super.onBackPressed()
        }
    }

    override fun onNavigateUp(): Boolean {
        val shard = pageHost.shard
        return if (shard is NavInterface && shard.onNavigateUp()) {
            true
        } else {
            super.onNavigateUp()
        }
    }
}
